File: src/cmcandy/C_Language_Answers/_0094.c

#include <stdio.h>
#include <stdlib.h>
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};


void getTreeSize(struct TreeNode* root ,int* returnSize){
    if(root==NULL){
        (*returnSize) = 0;
        return;
    }
    if(root->left!=NULL) {
        (*returnSize)++;
        getTreeSize(root->left,returnSize);
    }
    if(root->right!=NULL) {
        (*returnSize)++;
        getTreeSize(root->right,returnSize);
    }
    return ;
}

int getVal(struct TreeNode* root,int* res,int index){

    if(root->left!=NULL) index = getVal(root->left,res,index);
    res[index++] = root->val;
    if(root->right!=NULL) index = getVal(root->right,res,index);
    return index;

}
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* inorderTraversal(struct TreeNode* root, int* returnSize){
    (*returnSize) = 1;
    getTreeSize(root,returnSize);
    int* res = (int*) malloc(sizeof(int)*(*returnSize));
    if((*returnSize) ==0) return res;
    
    getVal(root,res,0);
    return res;
}

File: src/cmcandy/java_language_answers/_0094.java
package cmcandy.java_language_answers;

import java.util.*;
/**
 * Created by CHEN on 2020/6/3.
 */
public class _0094 {
    /**
     * Definition for a binary tree node. public class TreeNode { int val; TreeNode
     * left; TreeNode right; TreeNode(int x) { val = x; } }
     */
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    static class Solution {
        private List<Integer> res = new LinkedList<>();
        private Stack<TreeNode> stack = new Stack<>();

        public List<Integer> inorderTraversal(TreeNode root) {
            //  递归解法
            // if (root==null) return res;
            // if(root.left!=null) inorderTraversal(root.left);
            // res.add(root.val);
            // if(root.right!=null) inorderTraversal(root.right);
            // return res;

            //迭代解法
            TreeNode cur = root;
            while (cur != null || !stack.isEmpty()) {
                if (cur != null) {
                    stack.push(cur);
                    cur = cur.left;
                } else {
                    cur = stack.pop();
                    res.add(cur.val);
                    cur = cur.right;
                }
            }
            return res;
        }
    }
}